### PR TITLE
StringScanner#scan(Int), #skip(Int), #current_char, etc, and significantly speed up operations on multibyte characters

### PR DESCRIPTION
**Pull request by jneen**
Supersedes #11785, addresses most of #11259

This includes the following non-breaking changes to StringScanner:

* Refactor `#peek` to trust that `@byte_offset` is a valid char index, and use `Char::Reader` to read a specified number of characters ahead. The previous implementation would cause at least two full scans from the beginning of the string if it was not single byte optimizable, which could result in O(n^2) parses if `#peek` is used during a hot loop.
* Add overloads `#scan(Int)` and `#skip(Int)` to advance a specified number of characters, again using Char::Reader for non-optimizable strings. It is also a much more performant alternative to `scanner.offset += n`, which can cause two full string scans, both to retrieve the offset, and find the target offset.
* ~~Adds `#read_char : Char?` to scan forward by one (possibly multibyte) character and return it, or nil if at the end of the string. This is equivalent to Ruby's `#getch`, but the method name was chosen to be closer to IO.~~ Removed in favour of `#scan(1)`
* Adds `#current_char` and `#previous_char`, as well as UInt8-typed `#current_byte` and `#previous_byte`, which do not modify any state.
* Adds `#rewind(Int)` and `#peek_behind(Int)` to rewind the head or scan backwards by a specified number of chars, using `Char::Reader#previous_char` to find the byte offset in the presence of multibyte chars.
* Adds `#beginning_of_line?` (checks backwards one char for a newline, same as the Ruby impelementation) and `#unscan`, which rewinds the bytesize of the last match.
* Adds `#matched? : Bool` to check the presence of `@last_match`, similar to Ruby's implementation.
* Adjusts `#inspect` to show where the scan head is.

### CODE REVIEW

**Comment by HertzDevil** [comment]
Related: #11259

**Comment by jneen** [comment]
Happy to add those in this PR as well tbh, I don't think it'd be that difficult

**Comment by jneen** [comment]
Actually, ideally we'd be able to get an encoded character from the byte offset without scanning the whole string for multibyte chars, which `String#[]` (used by `#peek`) is wont to do. May require some extra functionality on String to be performant, I'll look into it.

**Comment by jneen** [comment]
...in fact, now that I look at it, it might be better for StringScanner to be backed by Char::Reader instead of String

File: src/string_scanner.cr [hunk 1]
```diff
@@ -359,7 +393,63 @@ class StringScanner
...
+
+  # :ditto:
+  def previous_char : Char
+    char_reader.previous_char
+  end
+
+  @[AlwaysInline]
+  private def char_reader : Char::Reader
+    Char::Reader.new(@str, @byte_offset)
   end
```
Comment (Sija) [inline src/string_scanner.cr]: Suggestion: use `delegate` macro.

```suggestion
  delegate :current_char?, :current_char, :previous_char?, :previous_char,
    to: char_reader

  private def char_reader : Char::Reader
    Char::Reader.new(@str, @byte_offset)
  end
```

Comment (jneen) [reply src/string_scanner.cr]: Hm, it certainly is shorter. Would that have an affect on the generated documentation?

Comment (Sija) [reply src/string_scanner.cr]: Negative, unfortunately - i.e. there is none.

Comment (jneen) [reply src/string_scanner.cr]: Alright, I think I'll stick with the more verbose form that gives generated docs, because I think they're important here.

**Comment by jneen** [comment]
Forgot to mention here, but this is ready for a review.

File: src/string_scanner.cr [hunk 2]
```diff
@@ -359,7 +389,63 @@ class StringScanner
   # Extracts a string corresponding to string[offset,*len*], without advancing
   # the scan offset.
   def peek(len) : String
-    @str[offset, len]
+    return "" if len.zero?
```
Comment (straight-shoota) [inline src/string_scanner.cr]: **thought:** I'm wondering if this optimization for empty string is even relevant? `lookahead_byte_length` and `String#byte_slice` already return early for zero length.
On the other hand, it doesn't hurt to have this explicit here 🤷 
Ditto below.

File: src/string_scanner.cr [hunk 3]
```diff
@@ -359,7 +407,63 @@ class StringScanner
...
+
+    @str.byte_slice(@byte_offset, lookahead_byte_length(len))
+  end
+
+  # Extracts a string corresponding to string[offset - len, len], without advancing
+  # the scan offset.
+  def peek_behind(len) : String
+    return "" if len.zero?
+    byte_len = lookbehind_byte_length(len)
+    byte_len = @byte_offset if byte_len > @byte_offset
```
Comment (straight-shoota) [inline src/string_scanner.cr]: **polish:** Simplify with `clamp`

```suggestion
    byte_len = byte_len.clamp(..@byte_offset)
```

File: src/string_scanner.cr [hunk 4]
```diff
@@ -359,7 +389,63 @@ class StringScanner
...
+  end
+
+  # :ditto:
+  def current_byte : UInt8
+    @str.byte_at(@byte_offset)
+  end
+
+  # Returns the byte before the scan head.
+  def previous_byte? : UInt8?
+    @str.byte_at?(@byte_offset - 1)
```
Comment (straight-shoota) [inline src/string_scanner.cr:469]: **issue:** This method returns the _last_ byte of the string when `@byte_offset == 0` because negative indices count from the back in `String#byte_at`. It should return `nil` instead.
This is an indicator that these new methods need more testing, especially for edge cases.

Ditto below.

File: src/string_scanner.cr [hunk 5]
```diff
@@ -359,7 +407,63 @@ class StringScanner
...
+  end
+
+  # :ditto:
+  def previous_char : Char
+    char_reader.previous_char
+  end
+
+  @[AlwaysInline]
+  private def char_reader : Char::Reader
+    Char::Reader.new(@str, @byte_offset)
```
Comment (straight-shoota) [inline src/string_scanner.cr]: **suggestion:** `@[AlwaysInline]` seems unnecessary here. It's a private method with a trivial body. The optimizer will certainly inline it anyway, even without the annotation. (And if not, might have a good reason to do so).

I don't think it's a good idea to sprinkle `@[AlwaysInline]` everywhere. In general, the optimizer should be expected to make the right choices.
Explicit annotations should be limited to cases where inlining is _required_ for semantic correctness, or when the optimizer might not recognize the proven potential of inlining. I don't think either is the case here.

```suggestion
  private def char_reader : Char::Reader
    Char::Reader.new(@str, @byte_offset)
```

File: spec/std/string_scanner_spec.cr [hunk 6]
```diff
@@ -302,20 +302,20 @@ describe StringScanner do
   describe "#inspect" do
     it "has information on the scanner" do
       s = StringScanner.new("this is a string")
-      s.inspect.should eq(%(#<StringScanner 0/16 "this " >))
+      s.inspect.should eq(%(#<StringScanner 0/16 |"this ">))
       s.scan(/\w+\s/)
-      s.inspect.should eq(%(#<StringScanner 5/16 "s is " >))
+      s.inspect.should eq(%(#<StringScanner 5/16 "s "|"is ">))
```
Comment (straight-shoota) [inline spec/std/string_scanner_spec.cr]: **thought:** Adding an indicator for the cursor position is great. But I'm not sure I like the additional quotes.
I suppose the intention is to avoid ambiguity if the string window contains the vertical bar character?
It also creates additional noise...

Maybe we could use a different character for the cursor. One that would be less likely to show up in a string?
For example, the broken bar `¦`, or the caret `‸`.

File: src/string_scanner.cr [hunk 7]
```diff
@@ -107,6 +120,30 @@ class StringScanner
...
+    scan_len = lookahead_byte_length(pattern)
+    idx = @byte_offset
+    @byte_offset += scan_len
+    result = @str.byte_slice(idx, scan_len)
+    @last_match = StringMatchData.new(result)
+    result
+  end
+
+  # Reads a single character from the string, or returns nil if at the end.
+  def read_char : Char?
```
Comment (straight-shoota) [inline src/string_scanner.cr]: **issue:** This method sticks out as a bit odd. It works very much like a `scan` method. But it has different semantics as it does not set `last_match`. Maybe it should?
Or is it even necessary to have this method? It should be equivalent to `scanner.current_char; scanner.skip(1)`.

Thinking about this, maybe an alternative or enhancement could be a `scan(Int)` method which consumes a fixed amount of characters. Then `read_char` would be equivalent to `scan(1)[0]`.

**Review by straight-shoota — COMMENTED** [review]
Tests are lacking and the class API docs need to be updated for the new features.
I'm also not entirely sure all of these additions make sense.
And we might want to split this big PR into separate ones for the individual changes.
But it should be okay to discuss everything here for now.

**Comment by jneen** [comment]
Happy to split it up! I just noticed the issue that @HertzDevil linked and figured I'd knock a few of those out while I was here.

File: src/string_scanner.cr (hunk 7 above)
Comment (jneen) [reply src/string_scanner.cr]: `#scan(Int)` and `#skip(Int)` are in fact part of this PR, so honestly this could be deleted without losing functionality.

File: spec/std/string_scanner_spec.cr (hunk 6 above)
Comment (jneen) [reply spec/std/string_scanner_spec.cr]: We can try that out for sure. I think the big thing is the output here might also have arbitrary whitespace including newlines or tabs, which is important not to hide. Perhaps we could use some kind of [combining diacritic](https://www.compart.com/en/unicode/block/U+0300)? We would have to trust that whatever this is being displayed on will render those properly though.

File: src/string_scanner.cr (hunk 5 above)
Comment (jneen) [reply src/string_scanner.cr]: No problem, I can change this. I mostly wanted to ensure that the struct only exists in local stack memory and isn't copied, but as you say the compiler will almost certainly do this anyways.

File: src/string_scanner.cr (hunk 4 above)
Comment (jneen) [reply src/string_scanner.cr:469]: Good catch, I'll fix and add a test.

File: src/string_scanner.cr (hunk 2 above)
Comment (jneen) [reply src/string_scanner.cr]: True. Mostly I wanted to be extra explicit and save a call to `byte_slice`, but as you've said it looks like `byte_slice` already applies this optimization, though with a trivial call to `Indexable.normalize_start_and_count`. All told this probably isn't necessary.

**Comment by jneen** [comment]
Question, which makes more sense for something like:

```crystal
# example A
s = StringScanner.new("foobar")
puts s.scan(1000)

# example B
s = StringScanner.new("foobar")
puts s.skip(1000)
```

Should this:
* Both be a no-op and return nil (equivalent to a non-matching regex or string)
* Both scan or skip to the end of the string and return the rest of the stream
* Scan is a no-op, but skip goes to the end of the string? In isolation this makes the most sense but put together is strange.


edit: Also for `#peek` and `#peek_behind`. I do feel like these ones should probably just return a best-effort String, which is what crystal currently does.

```crystal
s = StringScanner.new("foobar")
s.peek(1000) # => "foobar"
s.peek_behind(1000) # => ""
```

File: src/string_scanner.cr (hunk 2 above)
Comment (jneen) [reply src/string_scanner.cr]: It is deleted.

File: src/string_scanner.cr (hunk 4 above)
Comment (jneen) [reply src/string_scanner.cr:469]: Fixed and spec added.

File: src/string_scanner.cr (hunk 5 above)
Comment (jneen) [reply src/string_scanner.cr]: Removed and renamed to `make_char_reader` to emphasize that it's not a shared resource but a temporary utility struct.

File: src/string_scanner.cr (hunk 7 above)
Comment (jneen) [reply src/string_scanner.cr]: `#read_char` has been deleted in favour of `#scan(1).try(&.[0])` 👍

File: spec/std/string_scanner_spec.cr (hunk 6 above)
Comment (jneen) [reply spec/std/string_scanner_spec.cr]: After some experimentation I can confirm that combining diacritics are a bad idea here (they don't play well with String#inspect for one). Caret renders incorrectly in my terminal, being cut off at the bottom, but the [triangular bullet](https://www.compart.com/en/unicode/U+2023) `‣` seems alright, and even points at `current_char`. I've also added ellipsis characters where applicable and an `eos` indicator.

```crystal
    it "has information on the scanner" do
      s = StringScanner.new("this is a string")
      s.inspect.should eq(%(#<StringScanner 0/16 "‣this …">))
      s.scan(/\w+\s/)
      s.inspect.should eq(%(#<StringScanner 5/16 "…s ‣is …">))
      s.scan(/\w+\s/)
      s.inspect.should eq(%(#<StringScanner 8/16 "…s ‣a s…">))
      s.scan(/\w+\s\w+/)
      s.inspect.should eq(%(#<StringScanner 16/16 "…tring‣" eos>))
    end

    it "works with small strings" do
      s = StringScanner.new("hi")
      s.inspect.should eq(%(#<StringScanner 0/2 "‣hi">))
      s.scan(/\w/)
      s.inspect.should eq(%(#<StringScanner 1/2 "h‣i">))
      s.scan(/\w/)
      s.inspect.should eq(%(#<StringScanner 2/2 "hi‣" eos>))
    end
```

Comment (straight-shoota) [reply spec/std/string_scanner_spec.cr]: I would expect `¦` to have pretty high support.
A combining diacritic may be problematic when the string itself contains diacritics as well. Probably better to avoid that.
For escaping special characters, we can use `String#inspect_unquoted`.

**Comment by straight-shoota** [comment]
Yeah, whether the number of characters should be an exact match or upper limit is a tricky question. I would probably expect an exact match. Thus `scan(1000)` should be equivalent to `scan(/.{1000}/)`, not `scan(/.{0,1000}/)`.

Either way, both `scan` and `skip` should agree on this because they're closely related.

`peek` on the other hand is an entirely different operation, so aligning with that could be nice but it's not critical.

If there is a need for upper-limit matching, it could perhaps be implemented with a range: `scan(..1000)`. This is mostly an idea for the future, I don't think this is necessary right now.

**Comment by jneen** [comment]
I've edited the behaviour of `#scan(Int)` and `#skip(Int)` to both no-op and return nil if they go off the end of the string. I think this makes sense as you said, as well as considering that `#scan(1)` in a loop should return nil eventually. I also shored up the specs quite a bit and added missing failsafes on `#rewind`, which now just puts you back at the start if the length is too big.

As for documentation, I've added some doc comments - is there somewhere else I need to look to update docs?

File: spec/std/string_scanner_spec.cr (hunk 6 above)
Comment (jneen) [reply spec/std/string_scanner_spec.cr]: For sure. I think I slightly prefer `‣` over `¦` on account of it being less obtrusive and also pointing towards `current_char`, but I'm happy to change it. The current implementation does use `String#inspect_unquoted(IO)` and inserts the quotes manually. The quotes on the very outer part are probably important, as the segment may have spaces which are likely to be very difficult to see without at least the outer set of quotes.

As a preview, some options for comparison:

```
#<StringScanner 0/16 "‣this …">         #<StringScanner 0/16 "¦this …">
#<StringScanner 5/16 "…s ‣is …">        #<StringScanner 5/16 "…s ¦is …">
#<StringScanner 6/16 "… i‣s a…">        #<StringScanner 6/16 "… i¦s a…">
#<StringScanner 8/16 "…s ‣a s…">        #<StringScanner 8/16 "…s ¦a s…">
#<StringScanner 16/16 "…tring‣" eos>    #<StringScanner 16/16 "…tring¦" eos>

#<StringScanner 0/16 ‣this …>           #<StringScanner 0/16 ¦this …>
#<StringScanner 5/16 …s ‣is …>          #<StringScanner 5/16 …s ¦is …>
#<StringScanner 6/16 … i‣s a…>          #<StringScanner 6/16 … i¦s a…>
#<StringScanner 8/16 …s ‣a s…>          #<StringScanner 8/16 …s ¦a s…>
#<StringScanner 16/16 …tring‣ eos>      #<StringScanner 16/16 …tring¦ eos>
```

Comment (straight-shoota) [reply spec/std/string_scanner_spec.cr]: `‣` looks good 👍 
I had originally expected quotes would be great. But seeing the direct comparison when using `…` as indicators for "more" in that direction, I'm considering they might not be necessary and just add clutter. I don't think they're necessary for disambiguation.

I'm not sure if the EOS marker is very useful. If there is no trailing `…`, it already signals the end of the string. There could only be ambiguity if the string ends with `…`, but maybe that's acceptable? 🤷 The position indicator is already unambiguous.
If we want to keep that marker, I'd suggest to maybe use `\0` instead of `eos`?

Comment (jneen) [reply spec/std/string_scanner_spec.cr]: Yeah, actually now that we're using `#inspect_unquoted` on the string, leaving out the quotes seems fine. Here's an example at EOS, which I think is also fine, you can tell it's at EOS also because of the `16/16` part.

```
#<StringScanner 16/16 …tring‣>
```

That being said, there are cases where it breaks down, like:

```crystal
      s = StringScanner.new(" a ")
      s.skip(1)
      s.inspect.should eq(%(#<StringScanner 1/3  ‣a >))
```

Maybe if the inspected chunk contains spaces we use quotes?

edit: there's also the option of using a character to represent the beginning/ending of the string, to use when there are no ellipses, like:

```
#<StringScanner 0/16 ¦‣this …>
#<StringScanner 5/16 …s ‣is …>
#<StringScanner 6/16 … i‣s a…>
#<StringScanner 8/16 …s ‣a s…>
#<StringScanner 16/16 …tring‣¦>
#<StringScanner 0/2 ¦‣hi¦>
#<StringScanner 1/2 ¦h‣i¦>
#<StringScanner 2/2 ¦hi‣¦>
#<StringScanner 1/5 ¦ ‣a\n\t ¦>
#<StringScanner 0/9 ¦‣これは文字…>
#<StringScanner 1/9 ¦こ‣れは文字…>
#<StringScanner 2/9 ¦これ‣は文字…>
#<StringScanner 3/9 …れは‣文字列…>
#<StringScanner 6/9 …字列‣である¦>
#<StringScanner 9/9 …字列である‣¦>
```

Comment (jneen) [reply spec/std/string_scanner_spec.cr]: ~~...actually now that I look at it `inspect_unquoted` is still putting raw newlines and tabs and such into the output, a situation I think is likely to be common with writing parsers.~~ Nevermind, pebkac

Comment (straight-shoota) [reply spec/std/string_scanner_spec.cr]: How about using `"` instead of `¦`? That's a very common string delimiter.
```
#<StringScanner 0/16 "‣this …>
```

The idea would be that we're showing an excerpt of a long string literal. `…` just means the literal continues in that direction. But when we're at either end, the delimters are shown.

Comment (jneen) [reply spec/std/string_scanner_spec.cr]: Ya know, I thought I was going to hate having unbalanced quotes but the more I look at it the more it grows on me.  I'll put this in.

```
#<StringScanner 0/16 "‣this …>
#<StringScanner 5/16 …s ‣is …>
#<StringScanner 6/16 … i‣s a…>
#<StringScanner 8/16 …s ‣a s…>
#<StringScanner 16/16 …tring‣">
#<StringScanner 0/2 "‣hi">
#<StringScanner 1/2 "h‣i">
#<StringScanner 2/2 "hi‣">
#<StringScanner 1/5 " ‣a\n\t ">
#<StringScanner 0/9 "‣これは文字…>
#<StringScanner 1/9 "こ‣れは文字…>
#<StringScanner 2/9 "これ‣は文字…>
#<StringScanner 3/9 …れは‣文字列…>
#<StringScanner 6/9 …字列‣である">
#<StringScanner 9/9 …字列である‣">
```

**Comment by jneen** [comment]
Aside from commit gardening, I think this is ready for a re-review.

**Comment by jneen** [comment]
Rebased, commits gardened, specs green.

**Comment by jneen** [comment]
Rebased.

**Comment by jneen** [comment]
Rebased again.

File: src/string_scanner.cr [hunk 8]
```diff
@@ -41,33 +41,50 @@
...
+
   def initialize(@str : String)
     @byte_offset = 0
   end
 
   # Sets the *position* of the scan offset.
+  #
+  # **Note:** Moving the scan head with this method can cause performance issues in
+  # multibyte strings. For a more performant way to move the head, see
+  # [`#skip(Int)`](#skip%28pattern%3AInt%29%3AInt32%7CNil-instance-method) and `#rewind`.
```
Comment (Sija) [inline src/string_scanner.cr:87]: I'd avoid crystal docs-specific hacks.

```suggestion
  # `#skip(Int)` and `#rewind`.
```

File: src/string_scanner.cr [hunk 9]
```diff
@@ -78,9 +95,18 @@ class StringScanner
     @str.byte_index_to_char_index(@byte_offset).not_nil!
   end
 
+  # Rewinds the scan head by `len` characters.
```
Comment (Sija) [inline src/string_scanner.cr]:
```suggestion
  # Rewinds the scan head by *len* characters.
```

File: src/string_scanner.cr [hunk 10]
```diff
@@ -41,33 +41,50 @@
...
+  # it counts raw bytes instead of characters.
+  getter byte_offset : Int32
+
   def initialize(@str : String)
     @byte_offset = 0
   end
 
   # Sets the *position* of the scan offset.
+  #
+  # **Note:** Moving the scan head with this method can cause performance issues in
```
Comment (Sija) [inline src/string_scanner.cr]: There's a `NOTE` documentation admonition:

```suggestion
  # NOTE: Moving the scan head with this method can cause performance issues in
```

File: src/string_scanner.cr [hunk 11]
```diff
@@ -253,6 +318,14 @@ class StringScanner
     match(pattern, advance: false, anchored: true)
   end
 
+  # :ditto:
+  def check(pattern : Int) : String?
+    scan_len = lookahead_byte_length(pattern) || return nil
```
Comment (Sija) [inline src/string_scanner.cr:323]: Suggestion: prefer the usual early return style

```suggestion
    return unless scan_len = lookahead_byte_length(pattern)
```

File: src/string_scanner.cr [hunk 12]
```diff
@@ -356,10 +429,82 @@ class StringScanner
...
   end
 
-  # Extracts a string corresponding to string[offset,*len*], without advancing
-  # the scan offset.
+  # Extracts a string by looking ahead *len* characters, without advancing the
+  # scan offset. The return value has at most *len* characters, but may have fewer
+  # if the scan head is close to the end of the string.
   def peek(len) : String
-    @str[offset, len]
+    @str.byte_slice(@byte_offset, lookahead_byte_length(len) || @str.bytesize)
```
Comment (Sija) [inline src/string_scanner.cr]:
```suggestion
    byte_len = lookbehind_byte_length(len) || @str.bytesize
    @str.byte_slice(@byte_offset, byte_len)
```

File: src/string_scanner.cr [hunk 13]
```diff
@@ -356,10 +429,82 @@ class StringScanner
...
+  def current_char? : Char?
+    make_char_reader.current_char?
+  end
+
+  # Returns the character at the scan head, and errors if at the end. Does not
+  # move the scan head. This will properly decode the next character from the
+  # string, and may return a multi-byte character.
+  def current_char : Char
+    # [jneen] custom handling here because Char::Reader#current_char will just
+    # return '\0' at the end
```
Comment (Sija) [inline src/string_scanner.cr]: Is this comment still valid?

File: src/string_scanner.cr [hunk 14]
```diff
@@ -380,11 +525,124 @@ class StringScanner
...
+    end
+
+    # clamp in the case of small strings that don't have enough
+    # before *or* after
+    chars_before = chars_before.clamp(0, offset)
+    chars_after = chars_after.clamp(0, remaining)
+
+    indicator_char = '‣'
+    ellipsis_char = '…'
+    ending_char = '"'
```
Comment (Sija) [inline src/string_scanner.cr]: I'd suggest moving those into constants.

File: src/string_scanner.cr [hunk 15]
```diff
@@ -380,11 +525,124 @@ class StringScanner
...
+    raise ArgumentError.new("Negative lookahead count: #{len}") if len < 0
+    return 0 if len.zero?
+    if @str.single_byte_optimizable?
+      return len <= @str.bytesize - @byte_offset ? len : nil
+    end
+
+    # some redundant logic here from String#find_start_and_end, but in this case
+    # it is likely we are far into the string and len is small, so it is very
+    # important not to start at the beginning of the string.
+    reader = self.make_char_reader
```
Comment (Sija) [inline src/string_scanner.cr]: Redundant `self` receiver

```suggestion
    reader = make_char_reader
```

File: src/string_scanner.cr [hunk 16]
```diff
@@ -380,11 +525,124 @@ class StringScanner
...
+    return 0 if len.zero?
+
+    if @str.single_byte_optimizable?
+      return len <= @byte_offset ? len : nil
+    end
+
+    reader = self.make_char_reader
+
+    until len.zero?
+      reader.previous_char? || return nil
```
Comment (Sija) [inline src/string_scanner.cr:632]: Suggestion: prefer conventional early-return style.

```suggestion
      return unless reader.previous_char?
```

File: _typos.toml [hunk 17]
```diff
@@ -46,6 +46,8 @@ extend-ignore-re = [
   '"(aGFo|hel|Hel|thr|noe|Noe|BaR|fo|FO)', '(hel|Hel|worl|Worl|fo)"',
   '/fo|/FO', "'fo", "/\\(fo",
   'tr‸ue|fo‸o',
+  "#<StringScanner .*?>",
+  "eq\\(\".*?\"\\)",
```
Comment (Sija) [inline _typos.toml]: To avoid generalizing exceptions, perhaps it would be better to add `spec/std/string_scanner_spec.cr` to the `files.extend-exclude` array instead?

File: src/string_scanner.cr [hunk 18]
```diff
@@ -380,11 +525,124 @@ class StringScanner
...
+    # some redundant logic here from String#find_start_and_end, but in this case
+    # it is likely we are far into the string and len is small, so it is very
+    # important not to start at the beginning of the string.
+    reader = self.make_char_reader
+
+    current = reader.current_char?
+
+    loop do
+      break if len.zero?
+      return nil if current.nil?
```
Comment (Sija) [inline src/string_scanner.cr:608]: Suggestion: redundant `nil`

```suggestion
      return if current.nil?
```

File: src/string_scanner.cr (hunk 8 above)
Comment (jneen) [reply src/string_scanner.cr:87]: Is there another workaround? There doesn't seem to be any other way to get it to link to this implementation.

Comment (jneen) [reply src/string_scanner.cr:87]: (and the suggested change would link to the wrong implementation, cf https://github.com/crystal-lang/crystal/issues/16499)

File: src/string_scanner.cr (hunk 11 above)
Comment (jneen) [reply src/string_scanner.cr:323]: Is this really preferred? This is visually confusing to me as it is a common pattern with `==` in void methods, and would de-emphasize the possibility of nil.

File: src/string_scanner.cr (hunk 15 above)
Comment (jneen) [reply src/string_scanner.cr]: I understand that the semantics are the same, but I included it to make it visually distinct from, say, a variable. Is it really much clearer without the `self`?

File: _typos.toml (hunk 17 above)
Comment (jneen) [reply _typos.toml]: Is there a use case for tracking typos within test strings? Happy to move this out into something more specific if so. Also doesn't `extend-exclude` cause the *entire* file to be ignored, rather than just the test strings matching the regexp?

File: src/string_scanner.cr (hunk 18 above)
Comment (jneen) [reply src/string_scanner.cr:608]: Similarly to the above, I tend to prefer explicit nil return in methods that are not Nil-typed, even though `return` does default to nil. If this is established code style in the project I'm happy to change it but by default I tend to prefer explicitly calling out the places nil can occur.

File: src/string_scanner.cr (hunk 16 above)
Comment (jneen) [reply src/string_scanner.cr:632]: The reason I'm doing it this way is that `reader.previous_char?` has a side effect. The suggested change looks much more like a normal query.

File: src/string_scanner.cr (hunk 14 above)
Comment (jneen) [reply src/string_scanner.cr]: Done.

File: src/string_scanner.cr (hunk 10 above)
Comment (jneen) [reply src/string_scanner.cr]: Done, looks much better thank you!

File: src/string_scanner.cr (hunk 8 above)
Comment (ysbaddaden) [reply src/string_scanner.cr:87]: We shall fix the bug, then, not work our way around it with a hack that will silently break someday in the future.

File: src/string_scanner.cr (hunk 11 above)
Comment (straight-shoota) [reply src/string_scanner.cr:323]: We should avoid discussions about stylistic details in PR reviews. If there's no semantic difference, we should accept whatever stylistic choice the author makes. Or implement a style rule that affects the entire code base.

FTR: The suggestion is unacceptable to me because it hides the assignment to a local variable behind a trailing conditional, which negatively impacts readability.

File: _typos.toml (hunk 17 above)
Comment (ysbaddaden) [reply _typos.toml]: @jneen This is an interesting idea but it's unrelated to this issue. We shall discuss that in its own issue and have a distinct PR to introduce this change.

File: src/string_scanner.cr (hunk 11 above)
Comment (ysbaddaden) [reply src/string_scanner.cr:323]: This is the usual pattern, yes. Look around the stdlib to convince yourself.

The same is true of the other suggestions. They're the final polish, pedantic style changes and thus very boring, but the proposed changes make the changes fit better in the overall coding style :bow:

They also apply to other methods (such as `#unscan`): we don't have to `return nil` because `return` always returns `nil` by itself, we don't nil to check for `.nil?` because `if` already does it, and so on.

Crystal, like Ruby (and the Japanese language), is built on implicitness.

Comment (straight-shoota) [reply src/string_scanner.cr:323]: > but the proposed changes make the changes fit better in the overall coding style

If there is an established coding style, then yes, pedantic changes are acceptable and necessary. Ideally, this should be automated, of course.

For this specific suggestion, there is no explicit decision on either option. And I do not see convincing evidence that either is an established style in the code base.
There are plenty of instances of this exact pattern across stdlib. If we want to discourage it, we should address that existing code first. Then we can refer to that decision when adding new code.

Please, let's keep PR discussions free of stylistic dither. We've realized a long time ago that style discussions in PRs are not helpful.

The code is fine as is. We can refactor it later when we decide to implement a style rule.
Stylistic choices should not hold up merging this PR.

File: src/string_scanner.cr (hunk 13 above)
Comment (jneen) [reply src/string_scanner.cr]: Yes, though I suppose it could be more clear - I am explaining why this doesn't just call `make_char_reader.current_char`, since we'd like to raise IndexError when out of bounds instead of returning a null char.

File: _typos.toml (hunk 17 above)
Comment (jneen) [reply _typos.toml]: Understood. However, without this change this PR doesn't pass the typos check - what do you suggest I do in the meanwhile?

File: src/string_scanner.cr (hunk 12 above)
Comment (jneen) [reply src/string_scanner.cr]: I'll make this change. It seemed simple enough to me but may be a bit intense for someone coming across this code with no context.

File: _typos.toml (hunk 17 above)
Comment (straight-shoota) [reply _typos.toml]: It might bet better to just exclude the affected file instead of introducing regular expressions that try to match Crystal code.

Comment (jneen) [reply _typos.toml]: Okay! I presumed that excluding the entire file would be a touch heavy-handed but I'm happy to do that.

Comment (jneen) [reply _typos.toml]: Done.

File: src/string_scanner.cr (hunk 8 above)
Comment (jneen) [reply src/string_scanner.cr:87]: In the meanwhile is it okay to link to the wrong method here? That seems like a confusing experience for users - and that bug doesn't look simple or easy, in fact there's a solid argument that it's not a bug but rather an unimplemented feature.

File: src/string_scanner.cr (hunk 9 above)
Comment (jneen) [reply src/string_scanner.cr]: Done.

File: src/string_scanner.cr (hunk 15 above)
Comment (jneen) [reply src/string_scanner.cr]: While I think this looks more clearly like a method call and less like a variable name with `self`, with `make_` in the method name I don't think I feel strongly enough about it to protest. This is updated.

**Comment by jneen** [comment]
I have implemented many of the points of stylistic feedback here, with a few exceptions that I feel fairly strongly about:

* The docs should link to the correct method, the bug/unimplemented feature of linking to specific overloads notwithstanding. I think it's a confusing experience for the user to be linked to the wrong place, even if we plan to fix the issue someday. If we expect that fix/feature to land before this makes it into a release though, I agree it's reasonable to have the docs be incorrect until that point.
* In methods that are non-void - i.e. are normally expected to return a non-nil value, I strongly prefer an explicit `return nil` to make it visually obvious where `nil` comes from. Eliding it feels a bit overly clever to me, and undermines the communication intent involved in this kind of stylistic choice. If a method is *only* or even *primarily* used for its side-effects, I do agree that plain return is clearer. I should note that this style is also in use throughout the codebase, including in Char::Reader for a very similar use case:

```crystal
    @[AlwaysInline]
    private def decode_previous_char
      return nil if @pos == 0

      decode_char_before(@pos) do |code_point, width, error|
        @current_char_width = width
        @pos -= width
        @error = error
        @current_char = code_point.unsafe_chr
      end
    end
```

File: src/string_scanner.cr (hunk 13 above)
Comment (jneen) [reply src/string_scanner.cr]: Reworded the comment:

```crystal
  def current_char : Char
    # [jneen] Using the nilable version here and manually raising instead of
    # the perhaps more obvious `make_char_reader.current_char`. This is
    # because we want to raise an IndexError at the end of the stream, but that
    # method would return '\0' and not raise any error.
    current_char? || raise IndexError.new
  end
```

**Comment by jneen** [comment]
Perhaps some more substantial things to review, in case it's hard to find here - I do recognize that it's a fairly hefty PR:

* Does the new `StringScanner#inspect` output make sense? Is it an improvement over the current version?
* Does it make sense to leave `#offset` and `#offset=` as is, when they can potentially cause a performance footgun - one that will be undetectable until the user specifically tests on large strings containing multibyte characters?
* Are my implementations of `#lookahead_byte_length` and `#lookbehind_byte_length` robust and/or clear? Am I missing important test cases? Is this the right approach?
* Are there performance or consistency considerations I've overlooked? I do know this code more strongly trusts that `@byte_offset` is a valid character index than previously, in order to avoid re-scanning the full string - are there cases where using the API as designed would result in an invalid character index? Is that something we need to be concerned about?
* Are there important missing methods here that should be included? Are there methods here that are unnecessary and should be removed or split out?
* Do we care about handling other encodings without converting? I suppose that's a *much* bigger question though, I'm not sure even our Regex implementation does that at the moment. If that is ever planned, then Char::Reader will have to be substantially rewritten, and specific handlings for constant-time optimizable encodings should be implemented here.
* Have I inadvertently changed the behaviour of any existing method (I don't think that I have, but maybe I missed something)? Ideally this would be non-breaking on existing code (monkeypatches notwithstanding)

**Comment by straight-shoota** [comment]
Perhaps at this point it might be a good idea to split out some parts of this huge patch into smaller PRs.

In particular, I'd ask to have a separate PR for every change to existing functionality (improve perf for `#peek`, change format for `#inspect`). Modifications should be isolated to make it easier to track these changes.
Adding new functionality can be a bigger batch (or a set of batches).

Since these modifications depend on some of the new additions, we probably need multiple stages of pull requests. They should follow the commit history, and can batch new additions. I figure we could group it roughly like this:

1. Add `#scan(Int)`, `#skip(Int)`, `#check(Int)`
2. Add `#peek_behind` and improve perf of `#peek`
3. Improve format of `#inspect`
4. Add `current_char` etc.

The utility commits can be merged wherever they are needed first, or become their own PRs (e.g. the doc fixes).

Sorry for the bureaucracy 😢 But I believe this extra effort for a some smaller bites would help us a lot to review the changes and ease maintenance in the future.

**Comment by jneen** [comment]
That's alright with me - I'll open the first one today.

**Comment by jneen** [comment]
Closing in favour of a more piecemeal approach, beginning with #16557.
